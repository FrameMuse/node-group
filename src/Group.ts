import OrderedList from "./OrderedList"

interface CustomElementLifecycle {
  connectedCallback?(): void;
  disconnectedCallback?(): void;
  attributeChangedCallback?(name: string, oldValue: string | null, newValue: string | null): void;
}

/** @internal */
class GroupRelayElement extends HTMLElement implements CustomElementLifecycle {
  declare readonly group: Group

  connectedCallback() {
    this.hidden = true
    this.group.connectedCallback()
  }

  disconnectedCallback() {
    this.group.disconnectedCallback()
  }

  override remove(): void { this.group.remove() }
  override get textContent() { return this.group.textContent }

  static readonly TAG = "group-relay"
  static {
    if (window.customElements.get(GroupRelayElement.TAG) == null) {
      window.customElements.define(GroupRelayElement.TAG, GroupRelayElement)
    }
  }
}


// interface GroupOptions {
//   /**
//    * Ensures child nodes are always presented in the targeted parent.
//    */
//   lockChildren?: boolean
//   /**
//    * Pipes through child nodes to the targeted parent. Appending to the group
//    */
//   streamChildren?: boolean
// }

/**
 * A minimal version of document that stores a segment of a document structure just like `DocumentFragment`.
 * The key difference is that, it does affect the document and cause reflow when attached to a document,
 * while it does not have explicit wrapper like regular elements.
 * 
 * [DOM Proposal](https://github.com/whatwg/dom/issues/736)
 */
class Group extends DocumentFragment implements ChildNode {
  /** @internal */
  orderedNodes = new OrderedList<Node & Partial<ChildNode>>

  /**
   * Connection event relay element.
   * Initially added to the group.
   *
   * When group brings nodes to a parent, relay element relays connection callbacks and returns back to group.
   * 
   * @internal
   */
  relayElement = document.createElement(GroupRelayElement.TAG) as GroupRelayElement

  constructor() {
    super()

    super.appendChild(this.relayElement)
    // @ts-expect-error yes
    this.relayElement.group = this
  }

  /** @internal */
  connectedCallback() {
    // Append group nodes to targeted parent.
    this.relayElement.after(...this.orderedNodes)
    // Return tap.
    super.appendChild(this.relayElement)
  }

  /** @internal */
  disconnectedCallback() { }

  after(...nodes: (Node | string)[]): void {
    this.orderedNodes.getLast()?.after?.(...nodes)
  }
  before(...nodes: (Node | string)[]): void {
    this.orderedNodes.getFirst()?.before?.(...nodes)
  }
  remove(): void {
    for (const node of this.orderedNodes) {
      if (node == null) continue
      node.remove?.()
    }
  }
  replaceWith(...nodes: (Node | string)[]): void {
    this.after(...nodes)

    for (const node of this.orderedNodes) {
      if (node == null) continue

      nodes.includes(node)
      node.remove?.()
    }
  }


  override get children() {
    const elements = [...this.orderedNodes].filter(node => node instanceof HTMLElement)
    const collection: HTMLCollection = elements as never

    collection.item = NodeListItem
    collection.namedItem = CollectionNamedItem

    return collection
  }
  override get childNodes() {
    const nodes: ChildNode[] = [...this.orderedNodes] as ChildNode[]
    const nodeList: NodeListOf<ChildNode> = nodes as never

    nodeList.item = NodeListItem
    return nodeList
  }



  override appendChild<T extends Node>(node: T): T {
    if (node === this.relayElement as never) return node

    this.orderedNodes.append(node)
    this.after(node)
    return node
  }
  override removeChild<T extends Node>(child: T): T {
    this.orderedNodes.delete(child)
    return child
  }

  override replaceChild<T extends Node>(node: Node, child: T): T {
    this.orderedNodes.replaceItem(child, node)

    return child
  }
  override replaceChildren(...nodes: (Node | string)[]): void {
    if (nodes.every(node => this.orderedNodes.has(node))) return

    const oldNodesSnapshot = [...this.orderedNodes]

    this.append(...nodes)

    for (let i = 0; i < oldNodesSnapshot.length; i++) {
      const oldNode = oldNodesSnapshot[i]
      if (nodes.includes(oldNode)) continue

      oldNode.remove?.()
      this.orderedNodes.deleteAt(i)
    }
  }

  override append(...nodes: (Node | string)[]): void {
    const realNodes = nodes.map(createNode)

    this.after(...realNodes)
    this.orderedNodes.append(...realNodes)
  }
  override prepend(...nodes: (Node | string)[]): void {
    const realNodes = nodes.map(createNode)

    this.before(...realNodes)
    this.orderedNodes.prepend(...realNodes)
  }

  override insertBefore<T extends Node>(node: T, child: Node | null): T {
    if (child != null) {
      this.orderedNodes.insertBeforeItem(child, node)
    }
    return node
  }

  override contains(other: Node | null): boolean { return !!other && this.orderedNodes.has(other) }

  override get textContent() {
    let result = ""

    for (const node of this.orderedNodes) { result += node.textContent }

    return result
  }

  override get firstChild() {
    return (this.orderedNodes.getFirst() as ChildNode | undefined) ?? null
  }

  override get lastChild() {
    return (this.orderedNodes.getLast() as ChildNode | undefined) ?? null
  }

  override hasChildNodes(): boolean { return this.orderedNodes.length > 0 }

  // override getElementById(elementId: string): HTMLElement | null {}

  // static parentOf(child: ChildNode): Group | ParentNode | null {
  //   if (child instanceof)
  //     return child.parentNode
  // }
}

export default Group

function createNode(value: Node | string): Node {
  if (typeof value === "string") return new Text(value)
  return value
}



function NodeListItem<T>(this: Array<T>, index: number) {
  return this[index]
}

function CollectionNamedItem<T extends Element>(this: Array<T>, name: string) {
  return this.find(element => ((element as any).name || element.id) === name) ?? null
}