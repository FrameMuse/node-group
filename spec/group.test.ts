import "./dom"

// File: src/main.test.ts
// Bun >= v1.0 includes built-in test runner. No need for Jest/Mocha.
// Run: bun test
import { beforeEach, describe, expect, it } from "bun:test"

import Group from "../src/Group"

describe("Group behaviour", () => {
  beforeEach(() => {
    document.body.replaceChildren()
  })

  it("should behave like a moving document-fragment", () => {
    const parent1 = document.createElement("div")
    const parent2 = document.createElement("div")

    const group = new Group
    group.append("A", "B", "C")

    parent1.append(group)
    document.body.append(parent1)
    expect(parent1.textContent).toBe("ABC")

    parent2.append(group)
    document.body.append(parent2)
    expect(parent1.textContent).toBe("")
    expect(parent2.textContent).toBe("ABC")
  })

  it("should accept any mix of Node/strings", () => {
    const group = new Group

    const X = document.createTextNode("X")
    const Y = document.createElement("span")
    Y.textContent = "Y"

    group.append(X, Y, "Z")
    expect(group.textContent).toBe("XYZ")
  })

  it("should allow append *after* being attached", () => {
    const group = new Group

    const p = document.createElement("p")
    p.textContent = "Foo"

    group.append("Foo")

    const container = document.createElement("div")
    container.append(group)
    document.body.append(container)

    group.append("Bar", document.createTextNode("Baz"))

    expect(container.textContent).toBe("FooBarBaz")
    expect(group.childNodes.length).toBe(3)
  })

  it("should reflect DOM length", () => {
    const group = new Group
    group.append("1", "2")

    const c1 = document.createElement("ol")
    c1.append(group)
    document.body.append(c1)

    expect(c1.childNodes.length).toBe(2)
    expect(group.childNodes.length).toBe(2)

    const c2 = document.createElement("ol")
    c2.append(group)
    document.body.append(c2)

    expect(c1.childNodes.length).toBe(0) // Nodes were moved.
    expect(c2.childNodes.length).toBe(2)
    expect(group.childNodes.length).toBe(2)

  })

  it("should replace all children without losing remoteness", () => {
    const group = new Group
    group.append("Y")

    const container = document.createElement("div")
    container.append(group)
    document.body.append(container)

    group.replaceChildren("X")

    expect(group.textContent).toBe("X")
    expect(container.innerHTML).toBe("X")
  })
})

describe("Snapshot of document structure (advanced)", () => {
  it("snaps consistent node tree after sequential operations", () => {
    const group = new Group

    const p = document.createElement("p")
    group.append(p)
    group.append("X")
    group.append("?")
    group.append("Y")

    const container = document.createElement("div")
    container.append(group)
    document.body.append(container)

    const html = container.innerHTML
    expect(html).toMatchSnapshot()
  })
})
