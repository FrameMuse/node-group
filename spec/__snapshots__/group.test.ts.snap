// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot of document structure (advanced) snaps consistent node tree after sequential operations 1`] = `"<p></p>X?Y"`;
